### PR TITLE
Rewrite SQL antipattern with Gemini

### PR DESCRIPTION
**Pull request by franklinWhaite**
This is an experimental feature.

Adding a flag `--rewrite_sql` if used the output will include a rewritten SQL provided by Gemini.

Please note that the Gemini prompts are handled via a [yam file](https://github.com/GoogleCloudPlatform/bigquery-antipattern-recognition/blob/sqlAIrewrite/bigquery-antipattern-recognition/src/main/resources/antiPatternExamples.yaml).

This PR provides a the functionality to interact with Gemini and provide the user with a response. Fine tuning of the prompts is a subsequent step.

### CODE REVIEW

File: bigquery-antipattern-recognition/src/main/resources/antiPatternExamples.yaml [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    examples:
+    - "Input:\n select col1 from table1 where col2 like '%%asd%%' AND col3 = 'a';\n\
+      Output:\n select col1 from table1 where col3 = 'a' and col2 like '%%asd%%';\n\
+      Since the more restrictive equality filter is applied first then query will perform better"
+    - "Input:\n select col1 from table2 where col2 like '%%asd%%' AND col3 = 'a';\n\
+      Output:\n select col1 from table2 where col3 = 'a' and col2 like '%%asd%%';\n\
+      Since the more restrictive equality filter is applied first then query will perform better"
+  - name: StringComparison
+    description: | 
+      You now that in a predicate in a where string using a like is simpler and
```
Comment (afleisc) [inline bigquery-antipattern-recognition/src/main/resources/antiPatternExamples.yaml:17]: typo: now -> know

File: bigquery-antipattern-recognition/src/main/resources/antiPatternExamples.yaml [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    description: |
+      You now that ordering a large amount of rows has a considerable computational cost. 
+      Performance enhancements can be obtained be adding a LIMIT after the outermost order by
+    examples:
+    - "select col1 from table1 order by col1;\n\
+       Performance will increase by adding a limit\n\
+       select col1 from table1 order by col1 limit 1000;\n"
+  - name: LatestRecordWithAnalyticFun
+    description: |
+      You now that ROW_NUMBER() function is frequently used to get the latest record of a given partition.
```
Comment (afleisc) [inline bigquery-antipattern-recognition/src/main/resources/antiPatternExamples.yaml:67]: now -> know

File: bigquery-antipattern-recognition/src/main/resources/antiPatternExamples.yaml [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    examples:
+    - "select col1 from table1 where col2 = (select max(col3) from table2)\n\
+       it would be better to rewrite as follows:\n\
+       declare var1 [DATATYPE] DEFAULT (select 'a'); --add corresponding datatype here\n\
+       select col1 from table1 where col2 = var1;\n\
+       Note that since we do not know the datatype of the column used in the\n\ 
+       predicate we will leave a placeholder [dataype] so that somebody can manually replace the datatype\n"
+  - name: JoinOrder
+    description: |
+      You know that for optimal performance the first table in a JOIN should be te largest.
```
Comment (afleisc) [inline bigquery-antipattern-recognition/src/main/resources/antiPatternExamples.yaml:107]: te -> the

File: README.md [hunk 4]
```diff
@@ -54,6 +54,7 @@ CREATE OR REPLACE TABLE <my-project>.<my-dataset>.antipattern_output_table (
   query STRING,
   recommendation ARRAY<STRUCT<name STRING, description STRING>>,
   slot_hours FLOAT64,
+  optimized_sql STRING,
```
Comment (afleisc) [inline README.md:57]: need to add the optimized_sql field in the bq example doc as well ./EXAMPLES.md#bq-table---bq-table

**Review by afleisc — COMMENTED** [review]
I tried a bq-> example with 

`docker run -i bigquery-antipattern-recognition \
  --input_bq_table wpe-tam-sandbox-anthos.optimization_workshop.input_hash \
  --output_table wpe-tam-sandbox-anthos.optimization_workshop.antipattern_output_table_rec --rewrite_sql --processing_project_id wpe-tam-sandbox-anthos`
  
Screenshot 2024-03-04 at 9 47 21 AM

and got the following errors:

`2024-03-04 15:44:31:339 [main] INFO com.google.zetasql.toolkit.antipattern.rewriter.gemini.GeminiRewriter - sun.net.www.protocol.https.DelegateHttpsURLConnection:https://us-central1-aiplatform.googleapis.com/v1/projects/wpe-tam-sandbox-anthos/locations/us-central1/publishers/google/models/gemini-1.0-pro:streamGenerateContent?alt=sse
2024-03-04 15:44:31:342 [main] INFO com.google.zetasql.toolkit.antipattern.rewriter.gemini.GeminiRewriter - POST Response Code: 400
2024-03-04 15:44:31:343 [main] INFO com.google.zetasql.toolkit.antipattern.rewriter.gemini.GeminiRewriter - POST request failedBad Request
2024-03-04 15:44:31:343 [main] ERROR com.google.zetasql.toolkit.antipattern.rewriter.gemini.GeminiRewriter - Could not rewrite SQL. null`

**Comment by franklinWhaite** [comment]
@afleisc per the error "Could not rewrite SQL. null".

Is this a single error log or was the execution of the tool interrupted by the tool?

This could be expected behavior. Not al queries will be rewritable since Gemini is not expected to provide valid outputs 100% of times. The code should be able to log this as an error and continue with the execution for the remaining queries.

Is this the case?

**Comment by afleisc** [comment]
> @afleisc per the error "Could not rewrite SQL. null".
> 
> Is this a single error log or was the execution of the tool interrupted by the tool?
> 
> This could be expected behavior. Not al queries will be rewritable since Gemini is not expected to provide valid outputs 100% of times. The code should be able to log this as an error and continue with the execution for the remaining queries.
> 
> Is this the case?

Yes, the tool was not interrupted, it just produced no optimized version so expected behavior

**Review by afleisc — APPROVED** [review]

**Review by celiajmd — APPROVED** [review]
Reviewed changes. Tested in my project and the AI function works as described in Readme.

**Pull request merged**